misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/DebugOutputLogger.java

```diff
@@ -0,0 +1,108 @@
+
+/**
+ * Wrapper implementation for the DAP `client/output` endpoint, which can be used to send debug server and target VM
+ * outputs to the debugger client side.
+ *
+ * @since 2.0.0
+ */
+public class DebugOutputLogger {
+
+    private final IDebugProtocolClient client;
+
+    public DebugOutputLogger(IDebugProtocolClient client) {
+        this.client = client;
+    }
+
+    /**
+     * Processes messages which should treated as the program output.
+     *
+     * @param output string output produced in the debugger/target VM.
+     */
+    public void sendProgramOutput(String output) {
+        if (!isValid(output)) {
+            return;
+        }
+        if (!output.endsWith(System.lineSeparator())) {
+            output += System.lineSeparator();
+        }
+        OutputEventArguments outputEventArguments = new OutputEventArguments();
+        outputEventArguments.setOutput(output);
+        outputEventArguments.setCategory(OutputEventArgumentsCategory.STDOUT);
+        client.output(outputEventArguments);
+    }
+
+    /**
+     * Processes messages which should treated as the console output.
+     *
+     * @param output string output produced in the debugger/target VM.
+     */
+    public void sendConsoleOutput(String output) {
+        if (!isValid(output)) {
+            return;
+        }
+        if (!output.endsWith(System.lineSeparator())) {
+            output += System.lineSeparator();
+        }
+        OutputEventArguments outputArguments = new OutputEventArguments();
+        outputArguments.setOutput(output);
+        // Since Ballerina compiler logs and errors are redirected to the same stream (STDERR) by design, output
+        // category has to be derived based on the output prefix.
+        if (containsBalErrorPrefix(output)) {
+            outputArguments.setCategory(OutputEventArgumentsCategory.STDERR);
+        } else {
+            outputArguments.setCategory(OutputEventArgumentsCategory.CONSOLE);
+        }
+        client.output(outputArguments);
+    }
+
+    /**
+     * Processes messages which should treated as the error output.
+     *
+     * @param output string output produced in the debugger/target VM.
+     */
+    public void sendErrorOutput(String output) {
+        if (!isValid(output)) {
+            return;
+        }
+        if (!output.endsWith(System.lineSeparator())) {
+            output += System.lineSeparator();
+        }
+        OutputEventArguments outputEventArguments = new OutputEventArguments();
+        outputEventArguments.setOutput(output);
+        outputEventArguments.setCategory(OutputEventArgumentsCategory.STDERR);
+        client.output(outputEventArguments);
+    }
+
+    private static boolean containsBalErrorPrefix(String output) {
+        // Todo: Add more error patterns
+        return output.startsWith("ERROR") || output.startsWith("error:");
+    }
+
+    private static boolean isValid(String output) {
```

Comment:
The output of `isValid` method is always inverted in its usages. Therefore shall we change the method name to `isNotValid` and do the changes in the method body?

```
    private static boolean isNotValid(String output) {
        return output.startsWith("Listening for transport dt_socket")
            || output.startsWith("Please start the remote debugging client to continue")
            || output.startsWith("JAVACMD")
            || output.startsWith("Stream closed");
    }
```

Reply:
@praveennadarajah I prefer something like `isInternalLog` over `isNotValid` :) Changed in c823365. WDYT?

Reply:
+1